synapse/storage/data_stores/main/stream.py

```diff
@@ -797,18 +813,79 @@ def get_federation_out_pos(self, typ):
         return self.db.simple_select_one_onecol(
             table="federation_stream_position",
             retcol="stream_id",
-            keyvalues={"type": typ},
+            keyvalues={"type": typ, "instance_name": self._instance_name},
             desc="get_federation_out_pos",
         )
 
     def update_federation_out_pos(self, typ, stream_id):
         return self.db.simple_update_one(
             table="federation_stream_position",
-            keyvalues={"type": typ},
+            keyvalues={"type": typ, "instance_name": self._instance_name},
             updatevalues={"stream_id": stream_id},
             desc="update_federation_out_pos",
         )
 
+    def _reset_federation_positions_txn(self, txn):
+        """Fiddles with the `federation_stream_position` table to make it match
+        the configured federation sender instances during start up.
+        """
+
+        # The federation sender instances may have changed, so we need to
+        # massage the `federation_stream_position` table to have a row per type
+        # per instance sending federation. If there is a mismatch we update the
+        # table with the correct rows using the *minimum* stream ID seen. This
+        # may result in resending of events/EDUs to remote servers, but that is
+        # preferable to dropping them.
```

Comment:
Will this race if you start multiple of these at once?

I'm not 100% sure why we need to reset already existing ones back to the minimum. Is this due to the senders handling a different portion of destinations if the config changes?

Reply:
Hmm, you're right. Usually this isn't a problem because transactions will get retried, but that doesn't happen here. This is likely to happen in the wild when we restart workers at the same time. Will have a think

Reply:
> I'm not 100% sure why we need to reset already existing ones back to the minimum. Is this due to the senders handling a different portion of destinations if the config changes?

If the instances have changed then the destinations each worker handles gets rebalanced, which is why everything needs to start at the minimum

Reply:
> Hmm, you're right. Usually this isn't a problem because transactions will get retried, but that doesn't happen here. This is likely to happen in the wild when we restart workers at the same time. Will have a think

Of the changes left, I think this is the one that concerns me the most. The others are really aesthetic.

Reply:
Also -- what happens if only one worker restart (e.g. in a crash)? Will this be OK or will the running works get confused?

Reply:
I think the solution here is to change it so that this gets run the first time we try and get the federation stream position? That way the transactions will get correctly retried if there is a race (plus we can get rid of the hack to make it work with in memory DB). WDYT?

Reply:
I've done that

Reply:
I think the extra changes look sane.